lib/utils/gasEstimation.js

```diff
@@ -0,0 +1,279 @@
+const { BN } = require("ethereumjs-util");
+const RuntimeError = require("./runtimeerror");
+
+/**
+ * @param {Array} ops Array of all ops/steps of the VM
+ * @param {Number} lowestIndex Known lowestIndex
+ */
+module.exports = (vm, runArgs, err, callback) => {
+  const steps = stepTracker();
+
+  vm.on("step", steps.collect);
+
+  vm.runTx(runArgs, function(vmerr, result) {
+    const opsLength = steps.ops.length;
+    if (vmerr) {
+      // This is a check that has been in there for awhile. I'm unsure if it's required, but it can't hurt.
+      if (vmerr && vmerr instanceof Error === false) {
+        vmerr = new Error("VM error: " + vmerr);
+      }
+      // If we're given an error back directly, it's worse than a runtime error. Expose it and get out.
+      if (vmerr) {
+        return callback(vmerr, err);
+      }
+      // If no error, check for a runtime error. This can return null if no runtime error.
+      vmerr = RuntimeError.fromResults([runArgs.tx], { results: [result] });
+    } else if (!steps.success() || !opsLength || opsLength - 1 === steps.lowestGasIndex()) {
+      let estimate = parseInt(`0x${result.gasUsed.toString("hex")}`);
+      if (opsLength) {
+        const lastSVT = steps.lastSVT();
+        if (steps.success() && lastSVT > 0) {
+          const tailCost = sumRange(lastSVT, steps.ops.length - 1);
+          estimate += tailCost < 2300 ? 2300 - tailCost : 0;
+        }
+      }
+      result.gasEstimate = new BN(estimate);
+    } else {
+      // Base Fee 21000: cost of an elliptic curve operation to recover the sender pubkey/address from
+      // signature plus disk and bandwidth space of storing the transaction.
+      // Data Costs 4 * (# of 0 bytes) + 68 * (# of non-zero bytes)
+      const data = Array.from(runArgs.tx.data);
+      const costOfData = data.reduce((acc, curr) => acc + (curr === 0 ? 4 : 68), 21000);
```

Comment:
`// Data Costs 4 * (# of 0 bytes) + 68 * (# of non-zero bytes)`

So this line is just calculating the data costs ^.  Its sort of a very terse way of saying start with 21000, iterate through the data(each element represents a byte) and if the current byte is 0 add 4 otherwise add 68.